### PR TITLE
Models: add cue spec for text, news, and dashlist panels

### PR DESCRIPTION
**Pull request by ryantxu**
As a follow up to #32527, this converts some simple panels to the new cue format.  This is mostly an exercise to understand how the cue files can be applied to simple panels.

NOTE the generation is still hand crafted, but this should give an idea what the process may look like.

Failing with:

```
--- FAIL: TestScuemataBasics (0.60s)
    /home/ryan/workspace/grafana/grafana/pkg/schema/load/load_test.go:29: 
        	Error Trace:	load_test.go:29
        	Error:      	Received unexpected error:
        	            	required field PanelFieldConfig is optional in subsumed value (and 2 more errors)
        	Test:       	TestScuemataBasics
        	Messages:   	error while loading dist dashboard scuemata
```

### CODE REVIEW

File: public/app/plugins/panel/text/models.gen.ts
Comment (sdboyer) [inline public/app/plugins/panel/text/models.gen.ts:16]: My above-recommended change would change this to:

```ts
export interface PanelOptions {
  mode: "html" | "markdown";
  content: string;
} 
```


I don't actually remember if typescript will consider values from the `TextMode` enum to be to be structurally equivalent, and thus assignable here, which is obviously the goal. If not...mmm, yeah, need to think a bit about that one.

**Review by sdboyer — CHANGES_REQUESTED** [review]
I like that the very first follow-up PR has us hashing out important questions right away! :tada:

File: public/app/plugins/panel/text/models.cue
Comment (ryantxu) [reply public/app/plugins/panel/text/models.cue:3]: I made this as an example where the types we want to use are defined in the panel, not in grafana UI.

I'm open to whatever cue structure you think is best as long as we can generate a structure that matches what I proposed in `models.gen.ts` :)

Most plugins will have a mix of things defined in core with parameters defined locally

File: public/app/plugins/panel/text/models.gen.ts
Comment (ryantxu) [reply public/app/plugins/panel/text/models.gen.ts:16]: this works -- but would like to use this as a simple example of an explicit type enum.   (in typescript what you wrote and an enum defined with strings are actually identical)

**Comment by ryantxu** [comment]
@sdboyer can you take another look?  Also any advice on how to best debug validate syntax?  I installed a [vscode extension](https://github.com/cue-sh/vscode-cue), but that does not help too much.

These are simple panels that do not use FieldConfig (ie, they do not have custom extension for FieldConfig)

File: public/app/plugins/panel/text/models.cue
Comment (sdboyer) [inline public/app/plugins/panel/text/models.cue:15]: nit/q: this seems like something that's more the responsibility of e.g. a placeholder attribute on an input tag, no?

File: public/app/plugins/panel/text/models.gen.ts
Comment (sdboyer) [reply public/app/plugins/panel/text/models.gen.ts:16]: Mmmm...interesting. Yeah, we'll have to add support for this in cuetsy - effectively tell it to pop out a type with a particular name into the top-level scope, then reference it.

**Review by sdboyer — APPROVED** [review]
LGTM. Comments are just questions/nits.

File: public/app/plugins/panel/text/models.cue
Comment (ryantxu) [reply public/app/plugins/panel/text/models.cue:15]: Yes -- but in this case we actually save that default into the panel model and people start with editing it.  Typically "placeholder" text is replaced 100% once you start typing.

Not saying it is great... but matches current behavior

Comment (sdboyer) [reply public/app/plugins/panel/text/models.cue:15]: Another interesting case!

**Comment by sdboyer** [comment]
Oh, and i forgot to say - unfortunately, i haven't found much that provides direct editor support for working on CUE beyond that vscode extension. For my other CUE-related hacking, i've generally just set up poorman's validation - a `cue` command runs automatically in the background on save which ends up validating the appropriate files.

**Comment by sdboyer** [comment]
Oh, and i forgot to say - unfortunately, i haven't found much that provides direct editor support for working on CUE beyond that vscode extension. For my other CUE-related hacking, i've generally just set up poorman's validation - a `cue` command runs automatically in the background on save which ends up validating the appropriate files.

**Comment by torkelo** [comment]
@sdboyer @ryantxu anything blocking this PR?

**Comment by sdboyer** [comment]
Nothing from my end, it LGTM modulo the test failures

**Comment by ryantxu** [comment]
@dsotirakis -- I assigned this to you because it is a good example of the hurdle we need to pass before it is reasonable to ask people to write these cue files.  In the current state, the cue files *look* fine, but I have no idea what I actually have to fix to get this to pass:
```
--- FAIL: TestScuemataBasics (0.62s)
--
247 | load_test.go:29:
248 | Error Trace:	load_test.go:29
249 | Error:      	Received unexpected error:
250 | missing ',' in struct literal (and 1 more errors)
251 | Test:       	TestScuemataBasics
252 | Messages:   	error while loading dist dashboard scuemata
```

not sure where to go from there!

File: public/app/plugins/panel/dashlist/models.cue
Comment (dsotirakis) [inline public/app/plugins/panel/dashlist/models.cue:15]: Hello @ryantxu - sorry still stuck on CI stuff. With a first look it seems that these changes are needed IIUC? 
```suggestion
                    maxItems: int | *10
                    query: string | *""
                    folderId?: int
                    tags: [...string] | *[]
```
then there's `required field PanelFieldConfig is optional in subsumed value` which is another thing I guess - maybe something related to what Ying's doing?

Comment (sdboyer) [reply public/app/plugins/panel/dashlist/models.cue:15]: Yup, this change is correct. `[...string]` is how you express "a list of at least zero length with no maximum length, where all elements are a string".

This error: `required field PanelFieldConfig is optional in subsumed value` is a pretty classic example of a poor CUE error, and one we'll plausibly want to wrap.

It's saying that the schema being declared in this file - the object created starting on line 6 - is missing the required field `PanelFieldConfig`. That requirement is dictated by the `#PanelSchema` definition, [here](https://github.com/grafana/grafana/blob/master/cue/scuemata/panel-plugin.cue#L11), as we expect that all families in these files are instances of [`#PanelFamily`](https://github.com/grafana/grafana/blob/master/cue/scuemata/panel-plugin.cue#L18). Subsumption is mentioned because that's the [method](https://pkg.go.dev/cuelang.org/go@v0.4.0-alpha.1/cue#Value.Subsume) relied upon to check if the schema declared in this models.cue file is an instance of that `#PanelSchema`.

Comment (sdboyer) [reply public/app/plugins/panel/dashlist/models.cue:15]: More concrete, for absolute clarity. The error `required field PanelFieldConfig is optional in subsumed value` happens because this structure in this file:

```cue
            {
                PanelOptions: {
                    showStarred: bool | *true
                    showRecentlyViewed: bool | *false
                    showSearch: bool | *false
                    showHeadings: bool | *true
                    maxItems: int | *10;
                    query: string | *""
                    folderId?: int
                    tags: string[] | *[]
                }
            }
```

Does not contain a `PanelFieldConfig` field, which is specified as required by [`#PanelSchema`](https://github.com/grafana/grafana/blob/0d1cdbe227fe56077559f23f67c1a3f6785709eb/cue/scuemata/panel-plugin.cue#L9-L12):
```cue
#PanelSchema: {
    PanelOptions: {...}
    PanelFieldConfig: {...}
}
```

There are a few possible paths to addressing this, not all mutually exclusive:

* Improve the error message, as i noted in my previous comment.
* Make `PanelFieldConfig` an optional field in `#PanelSchema`.
* Expect that `models.cue` files import and unify the `#PanelFamily`; i.e., line 3 of this file becomes `Family: #PanelFamily & {`. Benefit might be that the file is more explicit about what's happening to the reader (assuming they understand CUE); cost is that it seems like a magical incantation to the file's author if they're inexperienced with CUE.
* Do the same unification as in the previous bullet, but do it on the fly in Go code. It'd have to be before [this call](https://github.com/grafana/grafana/blob/1336a57e99788e267e1ff882a33b005d66ad8b21/pkg/schema/load/panel.go#L142), which is where the error arises. More magical, but less confusing for folks inexperienced with CUE.

Comment (dsotirakis) [reply public/app/plugins/panel/dashlist/models.cue:15]: @ryantxu, @sdboyer just saw that Ryan made the field optional - is this the way to go in the end?

**Comment by dsotirakis** [comment]
Pushed some fixes - getting the error:

`field TextMode not allowed in closed struct (and 2 more errors)`, unless anyone knows I'll 🕵️

File: public/app/plugins/panel/dashlist/models.cue
Comment (ryantxu) [reply public/app/plugins/panel/dashlist/models.cue]: yes -- most panels do not need to define custom field config.

**Comment by ryantxu** [comment]
> (e.g., have TextMode defined somewhere in grafana-data?), IMO we should explore those options first.

um -- For things that *should* be shared across panels, that is fine, but it will be a mess to force everything there, and for external plugins it is impossible for them to define thier models in grafana-data

**Comment by sdboyer** [comment]
> um -- For things that should be shared across panels, that is fine, but it will be a mess to force everything there, and for external plugins it is impossible for them to define thier models in grafana-data

Sure, clear limitations on at least that as a strategy, got it. I'm just testing possibilities to see how far we can get with the minimum additions to what we have. (Basically, i'm triaging the deluge of stuff coming onto my plate.)

**Pull request merged**
